src/amt/APFProcessor.ts

```diff
@@ -249,193 +302,370 @@ const APFProcessor = {
+    parseIndex += authMethodLen
+
     let password: string = null
-    if (methodName === 'password') {
-      passwordLen = Common.ReadInt(data, 14 + usernameLen + serviceNameLen + methodNameLen)
-      password = data.substring(18 + usernameLen + serviceNameLen + methodNameLen, 18 + usernameLen + serviceNameLen + methodNameLen + passwordLen)
+    if (authMethod === 'password') {
+      // password plaintext ASCII encoding, maximum allowed size is 64 bytes.
+      if (dataLen < parseIndex + 5) {
+        return 0
+      }
+      // reserved byte
+      const reserved: number = data.charCodeAt(parseIndex)
+      parseIndex += 1
+      if (reserved !== 0) {
+        logger.warn(`${messages.MPS_WARN_RESERVED_VALUE}, ${reserved.toString()}`)
+      }
+      const passwordLen: number = Common.ReadInt(data, parseIndex)
+      parseIndex += 4
+      if (passwordLen > maxLen64) {
+        logger.warn(`${messages.MPS_ERR_MAXLEN} ${maxLen64} userAuthRequest:password`)
+        return -1
+      }
+      if (dataLen < parseIndex + passwordLen) {
+        return 0
+      }
+      password = data.substring(parseIndex, parseIndex + passwordLen)
+      parseIndex += passwordLen
+      logger.silly(`${messages.MPS_USER_AUTH_REQUEST} userAuthRequest:password has been parsed`)
     }
-    logger.silly(`${messages.MPS_USER_AUTH_REQUEST} usernameLen=${usernameLen} serviceNameLen=${serviceNameLen} methodNameLen=${methodNameLen}`)
-    logger.silly(`${messages.MPS_USER_AUTH_REQUEST} user=${username} service=${serviceName} method=${methodName}`)
+    logger.silly(`${messages.MPS_USER_AUTH_REQUEST} user=${user} service=${service} method=${authMethod}`)
 
     // Emit event to determine if user is authorized
     // TODO: verify this works correctly
-    APFProcessor.APFEvents.emit('userAuthRequest', socket, username, password)
+    APFProcessor.APFEvents.emit('userAuthRequest', socket, user, password)
 
-    return 18 + usernameLen + serviceNameLen + methodNameLen + passwordLen
+    return parseIndex
   },
 
-  protocolVersion: (socket: CIRASocket, length: number, data: string): number => {
-    if (length < 93) return 0
-    socket.tag.MajorVersion = Common.ReadInt(data, 1)
-    socket.tag.MinorVersion = Common.ReadInt(data, 5)
-    socket.tag.SystemId = APFProcessor.guidToStr(
-      Common.Rstr2hex(data.substring(13, 29))
-    ).toLowerCase()
-    logger.silly(
-      `${messages.MPS_PROTOCOL_VERSION}, ${socket.tag.MajorVersion}, ${socket.tag.MinorVersion}, ${socket.tag.SystemId}`
+  protocolVersion: (socket: CIRASocket, dataLen: number, data: string): number => {
+    if (dataLen < 93) return 0
+    let parseIndex = 1
+    const majorVersion = Common.ReadInt(data, parseIndex)
+    parseIndex += 4
+    const minorVersion = Common.ReadInt(data, parseIndex)
+    parseIndex += 4
+    // reserved 4 bytes
+    parseIndex += 4
+    const rstr = data.substring(parseIndex, parseIndex + 16)
+    parseIndex += 16
```

Comment:
For all of these functions, the parseIndex adds a lot more lines of code for not much increase in readability at the expense of unnecessary computations. I get that it lines up with the length values, but the previous values lined up with the offset values.